File: supermarket-pricing/typescript/src/Money.ts
/**
 * Monetary amount in cents. Integer arithmetic avoids floating-point rounding.
 */
export class Money {
  constructor(public readonly cents: number) {}

  static readonly zero = new Money(0);

  plus(other: Money): Money { return new Money(this.cents + other.cents); }
  minus(other: Money): Money { return new Money(this.cents - other.cents); }
  times(factor: number): Money { return new Money(this.cents * factor); }
  equals(other: Money): boolean { return this.cents === other.cents; }
}

File: supermarket-pricing/csharp/src/SupermarketPricing/Money.cs
namespace SupermarketPricing;

/// <summary>
/// Monetary amount in cents. Using integer cents avoids floating-point rounding
/// errors that plague price calculations. All arithmetic stays in whole cents;
/// conversions to display dollars happen at the boundary, not inside the domain.
/// </summary>
public readonly record struct Money(int Cents)
{
    public static Money Zero => new(0);
    public static Money operator +(Money a, Money b) => new(a.Cents + b.Cents);
    public static Money operator -(Money a, Money b) => new(a.Cents - b.Cents);
    public static Money operator *(Money a, int factor) => new(a.Cents * factor);
}
